src/bms/unparse.rs

```diff
@@ -0,0 +1,738 @@
+            }
+
+            for (obj_id, seek_time) in &others.seek_events {
+                tokens.push(Token::Seek(*obj_id, seek_time.clone()));
+            }
+
+            for extchr_event in &others.extchr_events {
+                tokens.push(Token::ExtChr(*extchr_event));
+            }
+        }
+
+        // Convert text definitions
+        for (obj_id, text) in &others.texts {
+            tokens.push(Token::Text(*obj_id, text));
+        }
+    }
+
+    // Helper methods for converting various data types to message format
+
+    fn obj_to_channel(&self, obj: &crate::bms::parse::model::obj::Obj) -> Channel {
+        match obj.kind {
+            NoteKind::Visible => Channel::Note {
+                kind: NoteKind::Visible,
+                side: obj.side,
+                key: obj.key,
+            },
+            NoteKind::Invisible => Channel::Note {
+                kind: NoteKind::Invisible,
+                side: obj.side,
+                key: obj.key,
+            },
+            NoteKind::Long => Channel::Note {
+                kind: NoteKind::Long,
+                side: obj.side,
+                key: obj.key,
+            },
+            NoteKind::Landmine => Channel::Note {
+                kind: NoteKind::Landmine,
+                side: obj.side,
+                key: obj.key,
+            },
+        }
+    }
+
+    fn obj_ids_to_message(&self, obj_ids: &[ObjId], _time: &ObjTime) -> String {
+        // This is a simplified implementation
+        // In a real implementation, you would need to handle the timing properly
+        if obj_ids.is_empty() {
+            return "00".to_string(); // Return empty object ID for no objects
+        }
+
+        obj_ids
+            .iter()
+            .map(|id| {
+                // Convert the string to a hex representation
+                id.to_string()
+            })
+            .collect()
+    }
+
+    fn obj_id_to_message(&self, obj_id: ObjId, _time: &ObjTime) -> String {
+        // This is a simplified implementation
+        // In a real implementation, you would need to handle the timing properly
+        obj_id.to_string()
+    }
+
+    fn bpm_to_message(&self, bpm: Decimal) -> String {
+        // Convert BPM to hex format
+        let bpm_u8 = bpm.to_u8().unwrap_or(120);
+        format!("{:02X}", bpm_u8)
+    }
+
+    fn stop_to_message(&self, duration: Decimal) -> String {
+        // Convert stop duration to hex format
+        let duration_u8 = duration.to_u8().unwrap_or(0);
+        format!("{:02X}", duration_u8)
+    }
+
+    fn scroll_to_message(&self, factor: Decimal) -> String {
+        // Convert scroll factor to hex format
+        let factor_u8 = factor.to_u8().unwrap_or(1);
+        format!("{:02X}", factor_u8)
+    }
+
+    fn speed_to_message(&self, factor: Decimal) -> String {
+        // Convert speed factor to hex format
+        let factor_u8 = factor.to_u8().unwrap_or(1);
+        format!("{:02X}", factor_u8)
+    }
```

Comment:
`#SCROLL` and `#SPEED` allow fractional value, so you should cast decimal into `f64` instead of `u8`.